Fix/log unavailable warning

jonasdebeukelaer opened the pull request:
* only allow troubleshoot link on error banner
* edit msg slightly when logs have gone away from cluster

Fixes #3711

jonasdebeukelaer commented:
feels like this failing build has nothing to do with ui? 🤔

Bobgy commented:
@jonasdebeukelaer Thanks a lot for your contribution! This is absolutely awesome for UX.
/lgtm
/approve
/ok-to-test

Bobgy commented:
Yes, the travis unit tests are flaky. I'll help you retry them.

Bobgy commented:
/retest

Bobgy commented:
@jonasdebeukelaer The error message for frontend unit test: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/3848/kubeflow-pipeline-frontend-test/1265449238280540160#1:build-log.txt%3A60

jiezhang commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -1091,9 +1091,10 @@ describe('RunDetails', () => {
             <Banner
               additionalInfo="getting logs failed"
-              message="Warning: failed to retrieve pod logs. Possible reasons include cluster autoscaling or pod preemption"
+              message="Failed to retrieve pod logs. Possible reasons include cluster autoscaling or pod preemption. Use Stackdriver Kubernetes Monitoring to view them."
```
Does it show the exclamation mark when mode is warning? Can we change the mode to info and rephrase the message? Something like "Failed to retrieve pod logs. Possible reasons include cluster autoscaling, pod garbage collection, and pod preemption. Use Stackdriver Kubernetes Monitoring to view them."

jonasdebeukelaer replied:
@jiezhang yes it does show a ⚠️ . Perhaps an info type banner does makes sense. I'll add one tomorrow

jonasdebeukelaer replied:
Actually on second thoughts maybe a warning makes sense since it could either be an error fetching the logs, or an info if it's just that a node as scaled down. Would it make sense to dig down deeper to know what has actually happened?

Bobgy replied:
Totally agree we should dig deeper on the error to know if it's expected. I'm not sure how that can be done, @jonasdebeukelaer would you want to investigate?

I think if we cannot know why, we need to keep the warning sign. The new sentence of "Use Stackdriver Kubernetes Monitoring to view them" is already a big step forward indicating users another option.

Maybe we should make the stackdriver link stand out when pod log cannot be achieved, what do you think?

jonasdebeukelaer replied:
sure, happy to have a look. 

So logic is:
* show logs if can
* if not
  * if node skipped - hide logs tab
   * if run is running/finished and pod can't be found - info banner saying that logs have gone w possible reasons, and optional 'access via stackdriver'
   * else error banner

and yes I can give the stackdriver link option more presence

Bobgy replied:
Thanks! This sounds perfect.

nit: `if run is running/finished and pod can't be found` -> `if run is in other states`, there are also like failed state that we should show log too

Bobgy replied:
Hmm, rethinking over it, I'm not sure we can get to the `else error banner` branch.

It is more of a problem that, when we get an error when fetching logs from pod, we need to parse the error message/status code and know some errors are expected (pod went away) and only show warning/error for others.

jiezhang replied:
NOTE that this happens for every pod - within hours of a pod's normal execution, it will be garbage collected. Why do we want to display a warning for 100% of the steps?

jonasdebeukelaer replied:
@jiezhang I've converted it to an info type banner, if a get to 'api/pod' returns a 404, so no more warnings. it will show an error if the logs don't display for a different reason though.

Ark-kun commented:
/test kubeflow-pipeline-frontend-test

jonasdebeukelaer commented:
changes:
* added info type banner
* use it if it's a case of pod not found by doing a get pod if errors, and seeing if it's a 404
* updated other banners in the side panel to use info type banner where appropriate
* added some missing unit tests

node: I've made namespace required for Apis.getPodLogs, not sure why it was optional before?

jonasdebeukelaer commented:
/retest

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -877,30 +905,52 @@ class RunDetails extends Page<RunDetailsInternalProps, RunDetailsState> {
+      logsBannerMessage = 'Failed to retrieve pod logs.';
+
+      // if pod can't be found, then assume it has gone from the server
+      if (await this.podNotFound(selectedNodeDetails.id, namespace)) {
```
nit: just confirm the error message returned from `Apis.getPodLogs` has no hint whether the pod is not found?
If no, this seems a reasonable workaround.

jonasdebeukelaer replied:
it does allude to workflow not being found, or something along those lines, but the error code is 500, so would have to do a msg contains text type check. Feels a bit brittle but happy to do that instead

Bobgy replied:
I'm okay with text check in this PR, if you feel you have capacity, we can also detect and change the error code in https://github.com/kubeflow/pipelines/blob/57f415d2d1960e4017d7eb05b0fea85d21436add/frontend/server/handlers/pod-logs.ts#L76

jonasdebeukelaer replied:
ohh I thought this was interfacing straight with argo 🤦 . This makes sense then yeah , will do here

jonasdebeukelaer replied:
so I've updated this but don't have time to stand up the whole backend to test it. is there some kind of short cut way to do it?

Bobgy replied:
Do you have a running KFP cluster?
If yes, you can follow https://github.com/kubeflow/pipelines/blob/master/frontend/README.md#proxy-to-a-real-cluster, you will need to develop "Client UI + Node server". It will send api requests to the cluster for anything else.

Bobgy replied:
unfortunately, we haven't added integration tests for pod logs api method.
https://github.com/kubeflow/pipelines/blob/d6920ca2ad0c36c195c8f483f4f1ac27d4b7e82f/frontend/server/app.test.ts#L1041-L1042

So I'd recommend just checking text match in this PR and you are welcome to change that implementation to node server status code in a separate one, preferably with test coverage.

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -877,30 +905,52 @@ class RunDetails extends Page<RunDetailsInternalProps, RunDetailsState> {
+      }
+
+      logsBannerAdditionalInfo += 'Error response: ' + errMsg;
+    } finally {
```
nit: maybe drop the finally? It seems unnecessary, because a normal code block after the try catch would suit our case too.

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -853,19 +864,36 @@ class RunDetails extends Page<RunDetailsInternalProps, RunDetailsState> {
+            sidepanelBannerMode = 'error';
+            break;
+          case NodePhase.FAILED:
+            sidepanelBannerMode = 'warning';
```
Sorry I don't have much context here, can you help me understand the difference between `ERROR` and `FAILED`?

jonasdebeukelaer replied:
I'll be honest I'm not 100% of this either, but I would understand this to be 
- ERROR if there was a error trying to set up the container or something, or network error, or issue in argo, and
- FAILED means the _execution_ of the container failed

Bobgy replied:
Thanks for the explanation, I'm not sure I understand why you want to show `warning` for failed node.
For users, it seems to me they should also perceive it as an error.

jonasdebeukelaer replied:
It's to try to separate pipelines failure from system failure, but happy to change to error type too

Bobgy replied:
Understood. IMO, can we add some wording in the error message to indicate difference between `fail` and `error`.
e.g. `error` -> `Pipelines System Error: original message`; while `fail` -> `Failed execution: ...`

what do you think?

This might be a little controversial, we can also revert this part keeping current behavior, so that we can get this PR merged. Opening a new PR for this specific issue.

jonasdebeukelaer replied:
Well I'll make em all Error type, and we can add a separate issue to add those prefixes?

Bobgy replied:
Sounds good! Thank you!

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -853,19 +864,36 @@ class RunDetails extends Page<RunDetailsInternalProps, RunDetailsState> {
     const workflow = this.state.workflow;
     const selectedNodeDetails = this.state.selectedNodeDetails;
+
+    let sidepanelBannerMode = 'warning' as Mode;
```
nit: `as` is a force type cast in typescript, recommend more idiomatic usage to specify type.
```
let sidepanelBannerMode: Mode = 'warning';
```

jonasdebeukelaer commented:
> Awesome! I've left some minor comments and nitpickings.

@Bobgy oops totally missed your update! Should have some time to fix this evening 👍

jonasdebeukelaer commented:
/retest

jonasdebeukelaer commented:
/retest

Bobgy commented on frontend/server/handlers/pod-logs.ts:
```diff
@@ -73,7 +73,13 @@ export function getPodLogsHandler(
       const stream = await getPodLogsStream(podName, podNamespace);
-      stream.on('error', err => res.status(500).send('Could not get main container logs: ' + err));
+      stream.on('error', err => {
+        if (err.message.indexOf('Unable to find pod log archive information') > -1) {
```
nit: `err?.message && err.message.indexOf`?
because someday maybe an err without message could be thrown

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -1110,7 +1150,7 @@ const VisualizationsTabContent: React.FC<{
         <>
           {viewerConfigs.length + generatedVisualizations.length === 0 && (
-            <Banner message='There are no visualizations in this step.' mode='warning' />
+            <Banner message='There are no visualizations in this step.' mode='info' />
```
Thanks! Makes sense

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -1212,12 +1259,12 @@ describe('RunDetails', () => {
         .at(1)
         .simulate('switch', STEP_TABS.LOGS);
       await getPodLogsSpy;
+      await getPodInfoSpy;
```
We no longer need this, right?

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -1198,9 +1219,35 @@ describe('RunDetails', () => {
+        .at(1)
+        .simulate('switch', STEP_TABS.LOGS);
+      await getPodLogsSpy;
+      await getPodInfoSpy;
```
ditto

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -1139,10 +1158,11 @@ describe('RunDetails', () => {
-              additionalInfo="getting logs failed"
-              message="Warning: failed to retrieve pod logs. Possible reasons include cluster autoscaling or pod preemption"
-              mode="warning"
+              additionalInfo="Possible reasons include cluster autoscaling, pod garbage collection and pod preemption. Error response: pod not found"
```
nit: let's order them based on likelihood?
pod garbage collection > cluster autoscaling ~= pod preemption, what do you think?

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -134,6 +137,7 @@ describe('RunDetails', () => {
     );
     isCustomVisualizationsAllowedSpy.mockImplementation(() => Promise.resolve(false));
     getPodLogsSpy.mockImplementation(() => 'test logs');
+    getPodInfoSpy.mockImplementation(() => ({ data: 'some data' } as JSONObject));
```
ditto: I think we no longer use it

Bobgy replied:
Ohh, thanks a lot! I missed it

Bobgy commented on frontend/src/components/Banner.test.tsx:
```diff
@@ -80,6 +104,29 @@ describe('Banner', () => {
+    const tree = shallow(
+      <Banner message='Some message' mode='warning' showTroubleshootingGuideLink={true} />,
+    );
+    expect(tree).toMatchInlineSnapshot(`
```
Same as above, please expect just the troubleshooting link doesn't show.
Warning mode snapshot was already tested above.

Bobgy commented on frontend/src/components/Banner.test.tsx:
```diff
@@ -52,8 +61,23 @@ describe('Banner', () => {
+        }}
+      />,
+    );
+    expect(tree.find('WithStyles(Button).refreshButton').at(0)).toMatchInlineSnapshot(`null`);
```
Hmmm, the selector is kind of hacky. What if RefreshButton later changed implementation, so it doesn't show as `WithStyles(Button)`.

Suggest finding the button by its text content: https://github.com/Bobgy/pipelines/blob/20157593093b888538095bdb2560c36b1354038e/frontend/src/components/NewRunParameters.test.tsx#L48-L52

also use `.toBeNull()` instead of match inline snapshot to be more explicit.

Bobgy commented on frontend/src/components/Banner.test.tsx:
```diff
@@ -35,6 +35,15 @@ describe('Banner', () => {
 
+  it('uses info mode when instructed', () => {
+    const tree = shallow(<Banner message={'Some message'} mode={'info'} />);
+    expect(tree.find('.icon').at(0)).toMatchInlineSnapshot(`
```
nit: maybe for info mode, you can match the full snapshot, because other modes all got a full snapshot?

Bobgy commented on frontend/src/components/Banner.test.tsx:
```diff
@@ -52,8 +57,23 @@ describe('Banner', () => {
+        }}
+      />,
+    );
+    expect(tree.findWhere(el => el.text() === 'Refresh')).toEqual({});
```
nit: I don't quite understand, why `toEqual({})`?
Let me take a look.

Bobgy replied:
Verified, `toEqual({})` will succeed no matter nodes found or not, because the wrapper helper object has no direct properties all the time.

Please use this way to verify not found:
https://github.com/kubeflow/pipelines/blob/fcd2559b2c846c2bd1e818cd72ebd8d64390258c/frontend/src/components/viewers/Tensorboard.test.tsx#L323-L330

jonasdebeukelaer replied:
oh thanks for that, sorry I'm quite new to ts tests

Bobgy replied:
No worries, you are doing great

Bobgy commented on frontend/server/handlers/pod-logs.ts:
```diff
@@ -73,7 +73,16 @@ export function getPodLogsHandler(
+          err?.message &&
+          err.message?.indexOf('Unable to find pod log archive information') > -1
+        ) {
+          res.status(404).send('pod not found');
```
nit: also append `err` to make sure we can debug if there are false-positives of 404s?

Bobgy replied:
ohhh, I forgot I approved before.
Feel free to update it in a following up PR
